src/Collections/Sequences/Actions.dfy

```diff
@@ -0,0 +1,261 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+/*******************************************************************************
+*  Original: Copyright (c) 2020 Secure Foundations Lab
+*  SPDX-License-Identifier: MIT
+*  
+*  Modifications and Extensions: Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT 
+*******************************************************************************/
+
+include "../../Wrappers.dfy"
+include "../../Action.dfy"
+include "Seq.dfy"
+
+module Actions {
+  import opened Wrappers
+  import opened A = Action
+  import opened Seq
+
+  method Map<A, R>(
+    action: Action<A, R>,
+    s: seq<A>
+  )
+    returns (res: seq<R>)
+    requires forall i | i in s :: action.Requires(i)
```

Comment:
(Just food for thought as I THINK we can address this later) - I wonder if we could make this less restrictive, so later elements could depend on earlier ones. Something like:

```
requires |s| > 0 ==> action.Requires(s[0])
requires forall i | 1 <= i < |s| :: 
  (exists r | action.Ensures(s[i - 1], r) ==> action.Requires(s[i]))
```

Reply:
This looks a bit suspicious.  Should it be parenthesized differently?

```
requires forall i | 1 <= i < |s| :: 
  ((exists r | action.Ensures(s[i - 1], r)) ==> action.Requires(s[i]))
```

Which is then equivalent to 

```
requires forall i | 1 <= i < |s| :: 
  (forall r :: action.Ensures(s[i - 1], r) ==> action.Requires(s[i]))
```

?